### ISSUE TITLE
Support cert based host auth

### ISSUE THREAD

**Issue by habnabit** [net-ssh/net-ssh#613]
### Expected behavior

When connecting to a host, if the host publishes only an `ssh-rsa` host key and `~/.ssh/known_hosts` contains an `ecdsa-sha2-nistp256` fingerprint for the host, OpenSSH will ignore the known host key and verify the `ssh-rsa` host key. In this case, despite the `ssh-rsa` key not being known, it is signed by a known host and considered trusted. I expected net-ssh to behave the same.

### Actual behavior

net-ssh does not trust the host, because it found _some_ known key, even if it's not of a key type the host is still publishing: https://github.com/net-ssh/net-ssh/blob/29a8e97ae8d2eaf674a8694b775dc119985fa5e9/lib/net/ssh/verifiers/always.rb#L17-L35

```
/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/verifiers/always.rb:46:in `process_cache_miss': fingerprint SHA256:[redacted] does not match for "[redacted]" (Net::SSH::HostKeyMismatch)
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/verifiers/always.rb:32:in `verify'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/verifiers/accept_new.rb:18:in `verify'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/verifiers/accept_new_or_local_tunnel.rb:17:in `verify'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/kex/diffie_hellman_group1_sha1.rb:188:in `verify_server_key'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/kex/diffie_hellman_group1_sha1.rb:72:in `exchange_keys'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/algorithms.rb:389:in `exchange_keys'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/algorithms.rb:211:in `proceed!'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/algorithms.rb:150:in `accept_kexinit'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:210:in `block in poll_message'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:190:in `loop'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:190:in `poll_message'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:225:in `block in wait'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:223:in `loop'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:223:in `wait'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh/transport/session.rb:90:in `initialize'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh.rb:242:in `new'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-5.0.2/lib/net/ssh.rb:242:in `start'
	/Users/habnabit/.rvm/gems/ruby-2.3.1/gems/net-ssh-gateway-2.0.0/lib/net/ssh/gateway.rb:57:in `initialize'
[...]
```

### System configuration

- net-ssh version 5.0.2 
- Ruby version ruby 2.3.1p112

### Next steps

I can submit a patch to fix this, but I'm not sure if or how net-ssh should decide to trust the host here. Is the right thing to consider the host unknown if the sets of host key algorithms are disjoint?

**Comment by mfazekas** [comment]
@habnabit Thanks much for the report.

I'm a bit confiused - are you using  `@cert-authority * < public key >` in `known_hosts`, so is this cert based host auth?

> In this case, despite the ssh-rsa key not being known, it is signed by a known host and considered trusted.

**Comment by habnabit** [comment]
@mfazekas yes, exactly. The precise fingerprint of the host's key is not known, but the host presents a certificate which is signed by a CA that _is_ in `known_hosts`.

**Comment by mfazekas** [comment]
@habnabit we don't recognize `@cert-authority` at the moment. We'd need to parse them and in verifiers we need to check if cert is signed by them, so we should treat them as `found`

**Comment by habnabit** [comment]
Okay, that makes sense. So, even if the host key's fingerprint isn't explicitly known, it should be treated as found and trusted if it's signed by a known and trusted CA.

**Comment by mfazekas** [comment]
> it should be treated as found and trusted if it's signed by a known and trusted CA.

Agree

**Comment by nabheet** [comment]
I just learned about the key signing feature in SSH and implemented it on our servers. It works with the SSH command line client - we no longer get prompted for key not being present in known_hosts. However, it was not working in our ruby code. And then I find this. I would love to help with this. Or at least investigate what it will take to get this feature.

Do you have any suggestions on how we could verify that the host key was signed by a specified CA?

**Comment by mfazekas** [comment]
@nabheet Looking at https://jameshfisher.com/2018/03/16/how-to-create-an-ssh-certificate-authority/
it looks that in known_hosts the format is `@cert-authority <server-pattern> <public_key>`
According to openssl docs you can use the method `verify(public_key)`.
See
https://ruby-doc.org/stdlib-2.5.1/libdoc/openssl/rdoc/OpenSSL/X509/Certificate.html

**Comment by aptituz** [comment]
Today I tried to come up with a PR for support of @cert-authority entries in known_hosts.

Eventually I came up with a working solution, but I figured that adding support for @cert-authority entries is not as easy as parsing them in the known_hosts class.

That's because those entries with a cert-authority marker need to be handled differently when verifying the certificate:

1. The type of the certificate-key is not relevant for the negotiation between server and client which key to request, because a CA key could be ssh-rsa while the server certificate could be (for example) ecdsa-sha2-nistp256-cert-v01@openssh.com
2. So when parsing known hosts we somehow need to track weither that entry is an actual host-key or a certificate.
3. In the negotiation of they keys we would need to prefer (all) the certificate algorithms, because we can't know for which we do have a certificate.
4. In the actual verification we would need to check the signature_key for the offered hostkey instead of the key itself.

Implementing this is actually quiet easy ... but it's a bit harder if the idea is to stay backwards compatible.

The naive implementation I did was to change keys_for to return a list of hashes with a type and the key instead of just a list of keys. But this practically breaks all callers of the KnownHosts class and thus breaks a lot of tests.

I tried to come up with a solution where parsing is done in an own method entries_for and keys_for returns only those entries which are host-keys (as it did before) ... but this feels wrong to me, as the negotiation-code and the verification-code would then need to evaluate two different lists and there is way more code to change in between.

I dunno what would be the preferred way to move forward. Therefore I'm writing this comment in the hope that the maintainers of Net::SSH might state their opinion on my ideas.

**Comment by aptituz** [comment]
Oh and btw. while testing my changes I learned that the `match` function in the known hosts class is not working correctly - at least not for certificates. Because the hostlist could be a single asterisk and the match-function wouldn't recognize that as it splits on asterisk.

A possible improvement would be:

```ruby
      def match(host, pattern)
        # see man 8 sshd for pattern details
        if pattern == '*'
          return true
        end

        pattern_regexp = pattern.split('*').map do |x|
          x.split('?').map do |y|
            Regexp.escape(y)
          end.join('.')
        end.join('[^.]*')

        host =~ Regexp.new("\\A#{pattern_regexp}\\z")
      end
```

**Comment by mfazekas** [comment]
@aptituz thanks much for looking into it.

I don't think that backward compatibility is a huge concern we can bump a major version.

Regarding keys_for i'm open to changing it, but first step can be keeping keys_for compatible and adding a new methods (entries_for), then migrating stuff over to entries_for.

>but this feels wrong to me, as the negotiation-code and the verification-code would then need to evaluate two different lists and there is way more code to change in between.

I'm not sure i fully understand this, if you have a PR it might be more clear to me.

Regarding `match` you seems to be right, a test case would be ideal.

**Comment by aptituz** [comment]
@mfazekas 

I stopped halfway implementing it that way, so I can't present you a PR. But it was about adding support for ca-keys with as few API breakage as possible and if that's not a big concern ... :)

I think for a clean implementation it would make sense to:

1. Introduce the suggested entries_for method, which will return known_host entries instead of host-keys (could be a KnownHostsEntry-class or a hash, but it should probably support three cases: host-keys, ca-keys and revoked keys)
2. Change host_keys to return entries with the type host as a first step towards that change as you suggested.
3. Change callers of keys_for to use entries_for and process type and key according to their meaning  in the specific context.

The third step would affect `search_for`, `search_in` and all other callers of keys_for (e.g. the test cases or direct callers in user-code).

Btw. on a somewhat related side-note: the code for negotiation / keying doesn't seem to support moving on to a different algorithm if the verification fails (as OpenSSH would do if there are multiple algorithms that could be agreed upon). Is that intended?

**Comment by coorasse** [comment]
argh! I guess we hit this very same issue, but I know too little to help with a PR.
I leave here for the record our error:
```
require 'net/ssh'
Net::SSH.start(HOST_URL)
```

raises

```
ecdh_sha2_nistp256.rb:88:in `send_kexinit': host key algorithm mismatch for signature 'ssh-ed25519' != 'ssh-ed25519-cert-v01@openssh.com' (Net::SSH::Exception)
```

**Comment by coorasse** [comment]
The problem is actually fixed in net-ssh 6.0. I don't have it anymore! 🎉 
Of course, I found out after 2 hours, but only 1 minute after leaving the previous message 😅

**Comment by carlhoerberg** [comment]
We're willing to pay for the development of this feature, please reach out to carl@84codes.com if that's interesting to anyone.
